src/crx.js

```diff
@@ -267,10 +267,14 @@ class ChromeExtension {
       throw new Error("No URL provided for update.xml.");
     }
 
+    var browserVersion = this.manifest.minimum_chrome_version
+      || (this.version < 3 && "29.0.0") // Earliest API available: https://developer.chrome.com/extensions/api_index
+      || "70.0.0"; // Around one year after Chromium started generating CRX3 packages
```

Comment:
Is there a source on this?

Reply:
@arkon Sorry, i had this info in my version of crx, but forgot to include it here. I added it in JSDoc now.

Basically my reasoning was that if Chromium switched to generating CRX3 packages on 16th October 2017, then Chrome should have started supporting it by 16th October 2018 (one year later, version 70.0.3538 according to Wikipedia: https://en.wikipedia.org/wiki/Google_Chrome_version_history). Well... probably ;).
I know for sure, that it generated CRX3 at the beginning of this year (and supported loading it too, of course).

As for CRX2, I think i made mistake: i forgot about CRX1. It is harder to find any info about it, but i found this:
https://medium.com/engineering-on-the-incline/distributing-chrome-extensions-for-older-versions-of-chrome-crx-version-number-invalid-be14325329e5
Thing is that article describes mandatory switch, not when v2 support started :(.

Chromium's file type sniffer has this:
https://chromium.googlesource.com/chromium/src/+/e435d6b7103f0d0c57133121454458bda6ccb69f/net/base/mime_sniffer.cc#454
Commit is from 2009-07-25. That adds sniffing of CRX2, so it may still not be officially supported yet. But Chrome 3.0.195 was released on 2009-10-12, and next one (4.0.249, 2010-1-25) mentions this:
> Extensions
(again, according to Wikipedia :).

On Chrome version history page, at version 27.0.1453, there is:
> Stop packaging Manifest version 1.0 extensions
which points to:
https://developer.chrome.com/extensions/manifestVersion#manifest-v1-support-schedule
But i'm afraid it's just about manifest version, not necessary CRX format version :(.

Anyway, i have no idea how much those formats differed, but i'm guessing (again, sorry ;), this module will not support CRX1 anyway (not much point in it now) and since Google says that earliest version with currently supported API is "29", then maybe it is enough?

Or we can set CRX2 to use "57.0.0" as default, as that was the version by which CRX2 became mandatory. At least according to the author of article i linked above.

The only sure way would be to run old versions of Chrome and test, but that would require too much time for me to spare on this, sorry.

Maybe "(almost) informed guesses" are enough for now? If they're incorrect, maybe some coder that has to support ancient browsers will issue a bug report :)?

Finally, that's just a `prodversionmin` number exported to XML file. It does not change CRX file. If developers are sure, that CRX works on earlier version, they can modify XML file or pass different value through manifest.json file. I added CLI argument in my module, but that's just because it also supports reading ZIP file from stdin (in which case it does not know data from manifest file). I could add it here too, i just thought it's not really necessary.

Reply:
I would've been satisfied with just the link in the JSDoc. :)

Seems like that commit was at [v64.0.3242.x](https://chromium.googlesource.com/chromium/src/+/b8bc9f99ef4ad6223dfdcafd924051561c05ac75/chrome/VERSION), so presumably ~v64/65 supported CRX3 files.

Also dug up [another commit](https://chromium.googlesource.com/chromium/src/+/5918d5fbdda9a61e2b83735975fcd7ebbf4d4645) related to CRX3 verification (v60.0.3108.x), so I guess they added it even earlier. That said, it at least appears that this version supported CRX2 or CRX3 files.

Reply:
Yeah, i saw that second commit too, but i thought that since they switched generator to CRX3 later, then maybe support for CRX3 was "flaky"/"test only" before that.

I did not think to check VERSION file :man_facepalming: , thanks :)!

16th October 2017 is a day before Chrome 62.0.3202, but i was guessing Chrome version gets Chromium stuff a bit later. That VERSION file shows 64.0.3242, Wikipedia says first "64.x" "published" was 64.0.3282 on 24th January 2018.
Should i change our CRX3 default from 70 to 64.0.3282?

Reply:
@arkon i found this: https://support.google.com/chrome/a/answer/7679408?hl=en
It contains this for Chrome 73:

> Starting with Chrome 75, all force-installed extensions will need to be packaged in the CRX3 format. Privately hosted extensions that were packaged using a custom script or a version of Chrome prior to Chrome 64.0.3242.0 must be repackaged. If your organization is force-installing privately hosted extensions packaged in CRX2 format and you don’t repackage them, they’ll stop updating in Chrome 75. And, new installations of the extension will fail.

First phrase is a bit confusing, because they already made it mandatory to use CRX3 in 73. But last phrase talks about 75 stopping updating CRX2 extensions, so maybe that's that.
Anyway, second phrase officially, indirectly confirms what we saw in repo history: 64.0.3242 started packaging CRX3 ;).

Reply:
I updated Wikipedia's Chrome version history page - thought it might be helpful to have CRX format info there too :).

Reply:
That link is referring to enterprise policies, which [this code review](https://chromium-review.googlesource.com/c/chromium/src/+/1297396) also refers to -- i.e. in the diff for `components/policy/resources/policy_templates.json`.

Reply:
I think it's safe to lower it to 64.0.3242.0 if it started packaging CRX3s at that point (since it would've have _had to_ accept them in that case).

Reply:
Yes, i know it was about enterprise policies, but it still confirmed version (unless they use different version numbers for enterprise?). They write about v75 because that's when force-installed extensions (AFAIK those are installed through configuration in system) will have to be CRX3, so my guess is that maybe CRX2 can still be force-installed for now, but i did not try :).

> I think it's safe to lower it to 64.0.3242.0 if it started packaging CRX3s at that point (since it would've have had to accept them in that case).

Yeah, i was thinking the same. Only i was not sure which Chrome version really switched to generating CRX3, so just to be on the safe side, i picked v70 ;). Since now we know the version, it's safe to switch.

Done. Thanks for help :).